File: src/main/scala/io/bespin/scala/mapreduce/bfs/IterateBfs.scala
package io.bespin.scala.mapreduce.bfs

import io.bespin.java.mapreduce.bfs.BfsNode
import io.bespin.scala.mapreduce.util.{BaseConfiguredTool, MapReduceSugar, TypedMapper, TypedReducer}
import org.apache.hadoop.io.IntWritable
import org.rogach.scallop.ScallopConf
import tl.lin.data.array.ArrayListOfIntsWritable
import tl.lin.data.map.HMapII

import scala.collection.JavaConverters._

/**
  * Tool for running one iteration of parallel breadth-first search.
  */
object IterateBfs extends BaseConfiguredTool with MapReduceSugar {
  private val Counter = "Counter"
  private val ReachableInMapper = "ReachableInMapper"
  private val ReachableInReducer = "ReachableInReducer"

  private object MapClass extends TypedMapper[IntWritable, BfsNode, IntWritable, BfsNode] {
    // For buffering distances keyed by destination node.
    private[this] val map = new HMapII

    // For passing along node structure.
    private[this] val intermediateStructure = new BfsNode

    override def setup(context: Context): Unit = {
      // Must clear the map for cases where mapper is
      // re-used (such as for local execution)
      map.clear()
    }

    override def map(nid: IntWritable, node: BfsNode, context: Context): Unit = {
      // Pass along node structure.
      intermediateStructure.setNodeId(node.getNodeId)
      intermediateStructure.setType(BfsNode.Type.Structure)
      intermediateStructure.setAdjacencyList(node.getAdjacencyList)

      context.write(nid, intermediateStructure)

      if(node.getDistance != Int.MaxValue) {
        context.getCounter(Counter, ReachableInMapper).increment(1)
        // Retain distance to self.
        map.put(nid.get, node.getDistance)

        val adj = node.getAdjacencyList
        val dist = node.getDistance + 1
        // Keep track of distance if it's shorter than previously
        // encountered, or if we haven't encountered this node.
        // Note that putting this into a more scala-idiomatic "foreach" seems
        // to cause it to silently skip nodes in some cases, so a while loop
        // is used instead.
        var i = 0
        while(i < adj.size) {
          val neighbor = adj.get(i)
          if(!map.containsKey(neighbor) || dist < map.get(neighbor)) {
            map.put(neighbor, dist)
          }
          i += 1
        }
      }
    }

    override def cleanup(context: Context): Unit = {
      val k = new IntWritable
      val dist = new BfsNode
      for (e <- map.entrySet.asScala) {
        k.set(e.getKey)
        dist.setNodeId(e.getKey)
        dist.setType(BfsNode.Type.Distance)
        dist.setDistance(e.getValue)
        context.write(k, dist)
      }
    }
  }

  private object ReduceClass extends TypedReducer[IntWritable, BfsNode, IntWritable, BfsNode] {
    private[this] val node = new BfsNode

    override def reduce(nid: IntWritable, iterable: Iterable[BfsNode], context: Context): Unit = {
      val values = iterable.iterator

      var structureReceived: Int = 0
      var dist: Int = Int.MaxValue

      while(values.hasNext) {
        val n = values.next

        if(n.getType == BfsNode.Type.Structure) {
          // This is the structure node; update accordingly
          node.setAdjacencyList(new ArrayListOfIntsWritable(n.getAdjacencyList))
          structureReceived += 1
        } else {
          // This is a message that contains distance
          dist = math.min(dist, n.getDistance)
        }
      }

      node.setType(BfsNode.Type.Complete)
      node.setNodeId(nid.get)
      node.setDistance(dist) // Update the final distance

      if(dist != Int.MaxValue) {
        context.getCounter(Counter, ReachableInReducer).increment(1)
      }

      // Error checking
      if(structureReceived == 1) {
        // Everything checks out, emit final node structures with updated distance.
        context.write(nid, node)
      } else if(structureReceived == 0) {
        // We get into this situation if there exists an edge pointing
        // to a node which has no corresponding node structure (i.e.,
        // distance was passed to a non-existent node)... log but move
        // on.
        log.warn("No structure received for nodeid: " + nid.get)
      } else {
        // This should never happen
        throw new RuntimeException("Multiple structure received for nodeid: " + nid.get
          + " struct: " + structureReceived)
      }
    }
  }

  private class Conf(args: Seq[String]) extends ScallopConf(args) {
    lazy val input = opt[String](descr = "input path", required = true)
    lazy val output = opt[String](descr = "output path", required = true)
    lazy val partitions = opt[Int](descr = "number of partitions", required = true)

    mainOptions = Seq(input, output, partitions)
  }

  override def run(argv: Array[String]): Int = {
    val args = new Conf(argv)

    log.info("Tool name: " + this.getClass.getSimpleName)
    log.info(" - input: " + args.input())
    log.info(" - output: " + args.output())
    log.info(" - partitions:" + args.partitions())

    val conf = getConf
    conf.set("mapred.child.java.opts", "-Xmx2048m")

    val thisJob = job(s"IterateBfs[input: ${args.input()}, output: ${args.output()}, " +
      s"partitions: ${args.partitions()}", conf)
      .sequenceFile[IntWritable, BfsNode](args.input())
      .map(MapClass)
      .reduce(ReduceClass, args.partitions())

    time {
      thisJob.saveAsSequenceFile(args.output(), deleteExisting = true)
    }

    0
  }
}

File: src/main/java/io/bespin/java/mapreduce/bfs/IterateBfs.java
/**
 * Bespin: reference implementations of "big data" algorithms
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.bespin.java.mapreduce.bfs;

import org.apache.hadoop.conf.Configured;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.input.SequenceFileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.apache.hadoop.mapreduce.lib.output.SequenceFileOutputFormat;
import org.apache.hadoop.util.Tool;
import org.apache.hadoop.util.ToolRunner;
import org.apache.log4j.Logger;
import org.kohsuke.args4j.CmdLineException;
import org.kohsuke.args4j.CmdLineParser;
import org.kohsuke.args4j.Option;
import org.kohsuke.args4j.ParserProperties;
import tl.lin.data.array.ArrayListOfInts;
import tl.lin.data.array.ArrayListOfIntsWritable;
import tl.lin.data.map.HMapII;
import tl.lin.data.map.MapII;

import java.io.IOException;
import java.util.Iterator;

/**
 * Tool for running one iteration of parallel breadth-first search.
 *
 * @author Jimmy Lin
 */
public class IterateBfs extends Configured implements Tool {
  private static final Logger LOG = Logger.getLogger(IterateBfs.class);

  private enum ReachableNodes {
    ReachableInMapper, ReachableInReducer
  }

  // Mapper with in-mapper combiner optimization.
  private static final class MapClass extends Mapper<IntWritable, BfsNode, IntWritable, BfsNode> {
    // For buffering distances keyed by destination node.
    private static final HMapII map = new HMapII();

    // For passing along node structure.
    private static final BfsNode intermediateStructure = new BfsNode();

    @Override
    public void map(IntWritable nid, BfsNode node, Context context)
        throws IOException, InterruptedException {
      // Pass along node structure.
      intermediateStructure.setNodeId(node.getNodeId());
      intermediateStructure.setType(BfsNode.Type.Structure);
      intermediateStructure.setAdjacencyList(node.getAdjacenyList());

      context.write(nid, intermediateStructure);

      if (node.getDistance() == Integer.MAX_VALUE) {
        return;
      }

      context.getCounter(ReachableNodes.ReachableInMapper).increment(1);
      // Retain distance to self.
      map.put(nid.get(), node.getDistance());

      ArrayListOfInts adj = node.getAdjacenyList();
      int dist = node.getDistance() + 1;
      // Keep track of shortest distance to neighbors.
      for (int i = 0; i < adj.size(); i++) {
        int neighbor = adj.get(i);

        // Keep track of distance if it's shorter than previously
        // encountered, or if we haven't encountered this node.
        if ((map.containsKey(neighbor) && dist < map.get(neighbor)) || !map.containsKey(neighbor)) {
          map.put(neighbor, dist);
        }
      }
    }

    @Override
    public void cleanup(Mapper<IntWritable, BfsNode, IntWritable, BfsNode>.Context context)
        throws IOException, InterruptedException {
      // Now emit the messages all at once.
      IntWritable k = new IntWritable();
      BfsNode dist = new BfsNode();

      for (MapII.Entry e : map.entrySet()) {
        k.set(e.getKey());

        dist.setNodeId(e.getKey());
        dist.setType(BfsNode.Type.Distance);
        dist.setDistance(e.getValue());

        context.write(k, dist);
      }
    }
  }

  // Reduce: select smallest of incoming distances, rewrite graph structure.
  private static final class ReduceClass extends Reducer<IntWritable, BfsNode, IntWritable, BfsNode> {
    private static final BfsNode node = new BfsNode();

    @Override
    public void reduce(IntWritable nid, Iterable<BfsNode> iterable, Context context)
        throws IOException, InterruptedException {

      Iterator<BfsNode> values = iterable.iterator();

      int structureReceived = 0;
      int dist = Integer.MAX_VALUE;
      while (values.hasNext()) {
        BfsNode n = values.next();

        if (n.getType() == BfsNode.Type.Structure) {
          // This is the structure; update accordingly.
          ArrayListOfIntsWritable list = n.getAdjacenyList();
          structureReceived++;

          int arr[] = new int[list.size()];
          for (int i = 0; i < list.size(); i++) {
            arr[i] = list.get(i);
          }

          node.setAdjacencyList(new ArrayListOfIntsWritable(arr));
        } else {
          // This is a message that contains distance.
          if (n.getDistance() < dist) {
            dist = n.getDistance();
          }
        }
      }

      node.setType(BfsNode.Type.Complete);
      node.setNodeId(nid.get());
      node.setDistance(dist); // Update the final distance.

      if (dist != Integer.MAX_VALUE) {
        context.getCounter(ReachableNodes.ReachableInReducer).increment(1);
      }

      // Error checking.
      if (structureReceived == 1) {
        // Everything checks out, emit final node structure with updated
        // distance.
        context.write(nid, node);
      } else if (structureReceived == 0) {
        // We get into this situation if there exists an edge pointing
        // to a node which has no corresponding node structure (i.e.,
        // distance was passed to a non-existent node)... log but move
        // on.
        LOG.warn("No structure received for nodeid: " + nid.get());
      } else {
        // This shouldn't happen!
        throw new RuntimeException("Multiple structure received for nodeid: " + nid.get()
            + " struct: " + structureReceived);
      }
    }
  }

  private IterateBfs() {}

  private static final class Args {
    @Option(name = "-input", metaVar = "[path]", required = true, usage = "input path")
    String input;

    @Option(name = "-output", metaVar = "[path]", required = true, usage = "output path")
    String output;

    @Option(name = "-partitions", metaVar = "[num]", required = true, usage = "number of partitions")
    int partitions;
  }

  /**
   * Runs this tool.
   */
  @Override
  public int run(String[] argv) throws Exception {
    final Args args = new Args();
    CmdLineParser parser = new CmdLineParser(args, ParserProperties.defaults().withUsageWidth(100));

    try {
      parser.parseArgument(argv);
    } catch (CmdLineException e) {
      System.err.println(e.getMessage());
      parser.printUsage(System.err);
      return -1;
    }

    LOG.info("Tool name: " + this.getClass().getName());
    LOG.info(" - input: " + args.input);
    LOG.info(" - output: " + args.output);
    LOG.info(" - partitions: " + args.partitions);

    getConf().set("mapred.child.java.opts", "-Xmx2048m");

    Job job = Job.getInstance(getConf());
    job.setJobName(String.format("IterateBfs[input: %s, output: %s, partitions: %d]",
        args.input, args.output, args.partitions));
    job.setJarByClass(EncodeBfsGraph.class);

    job.setNumReduceTasks(args.partitions);

    FileInputFormat.addInputPath(job, new Path(args.input));
    FileOutputFormat.setOutputPath(job, new Path(args.output));

    job.setInputFormatClass(SequenceFileInputFormat.class);
    job.setOutputFormatClass(SequenceFileOutputFormat.class);

    job.setMapOutputKeyClass(IntWritable.class);
    job.setMapOutputValueClass(BfsNode.class);
    job.setOutputKeyClass(IntWritable.class);
    job.setOutputValueClass(BfsNode.class);

    job.setMapperClass(MapClass.class);
    job.setReducerClass(ReduceClass.class);

    // Delete the output directory if it exists already.
    FileSystem.get(job.getConfiguration()).delete(new Path(args.output), true);

    job.waitForCompletion(true);

    return 0;
  }

  /**
   * Dispatches command-line arguments to the tool via the {@code ToolRunner}.
   *
   * @param args command-line arguments
   * @throws Exception if tool encounters an exception
   */
  public static void main(String[] args) throws Exception {
    int res = ToolRunner.run(new IterateBfs(), args);
    System.exit(res);
  }
}